File: Algorithm/Python/367. Valid Perfect Square.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-26 21:28:19
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-26 21:28:26
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        left, right = 1, num
        while left < right:
            mid = left + (right - left)/2
            area = mid*mid
            if area == num:
                return True
            elif area > num:
                right = mid -1
            else:
                left = mid + 1
        if left*left == num:
            return True
        return False

File: Algorithm/Java/367. Valid Perfect Square.java
/*
* @Author: WuLC
* @Date:   2017-08-09 15:54:53
* @Last Modified by:   WuLC
* @Last Modified time: 2017-08-09 15:55:19
* @Email: liangchaowu5@gmail.com
*/

// use long type instead of int
public class Solution
{
    public boolean isPerfectSquare(int num) 
    {
        long left = 1, right = num, mid, tmp;
        while (left <= right)
        {
            mid = left + ((right - left)>>1);
            tmp = mid * mid;
            if (tmp == num) return true;
            else if(tmp > num) right = mid - 1;
            else left = mid + 1;  
        }
        return false;
    }
}